### PR TITLE
Add "apply on update" option to default values

### PR DESCRIPTION
**Pull request by m-kuhn**
This is an early non-functional proposal to get feedback on the general API and workflow.

The idea is to have a new checkbox on the fields properties, next to the default value which says "apply on update". If this is activated, default values will be applied again when a feature is edited, a common usage for this is a "last edited" timestamp.

On the feature form, the evaluated value will be shown directly when the feature is edited (so it can be manually adjusted, unless the field is set to read only).

What's open for discussion is what to do for geometry edits. Should this go into `QgsVectorLayer::changeGeometry()` and `QgsVectorLayer::updateFeature()` so it will be done transparently in the background? It sounds like a safe approach, but I dislike to have this react on API calls whereas the field updates would be on the form logic. Happy for inputs here.

@nyalldawson especially wondering for your feedback on the [`QString defaultValue` to `QgsDefaultValue defaultValue`](https://github.com/qgis/QGIS/pull/5218/files#diff-3225c6a3b9805892152fca25f49f0c70) change. Should be fine to do, `QgsVectorLayer::defaultValueExpression`  has been added in 3.0 so we wouldn't even have an API break there.

### CODE REVIEW

**Comment by nyalldawson** [comment]
What you've got so far looks good to me!

Let me ponder on your ideas about where to do the updates...

**Comment by m-kuhn** [comment]
> Let me ponder on your ideas about where to do the updates...

Just thought about this again, it would be preferable to do all the updates directly on api calls even for attributes.

The reason is, that the attribute form should sometimes show the current (non-updated) value for inspection. This sometimes also is desirable even when in edit mode. On the attribute table form view where features get committed automatically when changing features this gets particularly difficult.

Does that sound ok for everyone?

**Comment by m-kuhn** [comment]
Idea on the "UX front": fields with auto-update will be read-only, the label (or a symbol next to it) yellow and a tooltip saying

> This field is updated automatically when making changes to this feature. Value preview: "[evaluated expression]".

**Comment by m-kuhn** [comment]
image

**Comment by m-kuhn** [comment]
Manually merged

File: src/core/qgsvectorlayer.cpp [hunk 1]
```diff
@@ -2270,33 +2296,40 @@ bool QgsVectorLayer::changeGeometry( QgsFeatureId fid, const QgsGeometry &geom )
...
-bool QgsVectorLayer::changeAttributeValue( QgsFeatureId fid, int field, const QVariant &newValue, const QVariant &oldValue )
+bool QgsVectorLayer::changeAttributeValue( QgsFeatureId fid, int field, const QVariant &newValue, const QVariant &oldValue, bool skipDefaultValues )
 {
+  bool result = false;
+
   switch ( fields().fieldOrigin( field ) )
   {
     case QgsFields::OriginJoin:
-      return mJoinBuffer->changeAttributeValue( fid, field, newValue, oldValue );
+      result = mJoinBuffer->changeAttributeValue( fid, field, newValue, oldValue );
```
Comment (nyalldawson) [inline src/core/qgsvectorlayer.cpp]: I think you're missing a "break" here (and also for the other cases modified here)

Comment (m-kuhn) [reply src/core/qgsvectorlayer.cpp]: right, thanks!

**Pull request closed without merging**